src/HealthChecks/HealthChecks/src/DefaultHealthCheckService.cs

```diff
@@ -39,76 +39,81 @@ internal class DefaultHealthCheckService : HealthCheckService
-                        try
-                        {
-                            var result = await healthCheck.CheckHealthAsync(context, cancellationToken);
-                            var duration = stopwatch.GetElapsedTime();
-
-                            entry = new HealthReportEntry(
-                                status: result.Status,
-                                description: result.Description,
-                                duration: duration,
-                                exception: result.Exception,
-                                data: result.Data);
-
-                            Log.HealthCheckEnd(_logger, registration, entry, duration);
-                            Log.HealthCheckData(_logger, registration, entry);
-                        }
-
-                        // Allow cancellation to propagate.
-                        catch (Exception ex) when (ex as OperationCanceledException == null)
-                        {
-                            var duration = stopwatch.GetElapsedTime();
-                            entry = new HealthReportEntry(
-                                status: HealthStatus.Unhealthy,
-                                description: ex.Message,
-                                duration: duration,
-                                exception: ex,
-                                data: null);
-
-                            Log.HealthCheckError(_logger, registration, ex, duration);
-                        }
-
-                        entries[registration.Name] = entry;
+                        var duration = stopwatch.GetElapsedTime();
+                        entry = new HealthReportEntry(
+                            status: HealthStatus.Unhealthy,
+                            description: ex.Message,
+                            duration: duration,
+                            exception: ex,
+                            data: null);
+
+                        Log.HealthCheckError(_logger, registration, ex, duration);
                     }
+
+                    return (registration.Name, entry);
                 }
+            }
+
+            IEnumerable<HealthCheckRegistration> registrations = _options.Value.Registrations;
+            if (predicate != null)
+            {
+                registrations = registrations.Where(predicate);
+            }
+
+            var totalTime = ValueStopwatch.StartNew();
+            Log.HealthCheckProcessingBegin(_logger);
 
-                var totalElapsedTime = totalTime.GetElapsedTime();
-                var report = new HealthReport(entries, totalElapsedTime);
-                Log.HealthCheckProcessingEnd(_logger, report.Status, totalElapsedTime);
-                return report;
+            (string registrationName, HealthReportEntry result)[] results;
+            using (var scope = _scopeFactory.CreateScope())
+            {
+                results = await Task.WhenAll(registrations.Select(r => RunCheckAsync(scope, r)));
```

Comment:
I'd like to avoid the capturing lambda and `.Select` allocation if possible - just use a `foreach`?

This also should have a `.ConfigureAwait(false)`.